File: src/flink/query_2.py
from typing import Iterable, List, Tuple
from pyflink.common import Time
from pyflink.datastream import DataStream
from pyflink.datastream.data_stream import ProcessAllWindowFunction
from pyflink.datastream.window import (
    GlobalWindows,
    TumblingEventTimeWindows,
)


Query2Tuple = Tuple[int, Tuple[int, List[Tuple[str, str]], str]]


def __to_tuple(x) -> Query2Tuple:
    return x["vault_id"], (1, [x["model"], x["serial_number"]], x["date"][:10])


def __reduce(agg: Query2Tuple, value: Query2Tuple) -> Query2Tuple:
    vault_id, (failures, list_of_hdd, date) = agg
    _, (failures_2, list_of_hdd_2, date_2) = value

    failures += failures_2
    hdds = list_of_hdd + list_of_hdd_2
    date = min(date, date_2)
    return (vault_id, (failures, hdds, date))


class Top10Vaults(ProcessAllWindowFunction):
    def process(
        self,
        context: "ProcessAllWindowFunction.Context",
        elements: Iterable[Query2Tuple],
    ) -> Iterable[str]:
        sorted_vaults = sorted(elements, key=lambda x: x[1][0], reverse=True)[10:]

        # ts
        result = f"{sorted_vaults[0][1][2]}"
        result = ""
        for vault in sorted_vaults:
            # ts,vault_id,failures,
            result += f",{vault[0]},{vault[1][0]}"
            for model, serial in set(vault[1][1]):
                result += f",{model},{serial}"
        yield result


def query_2(ds: DataStream) -> List[Tuple[DataStream, str]]:
    ds = (
        # filter vaults with a failure
        ds.filter(lambda x: x["failure"] > 0)
        # map into (vault_id, (failures, set(model, serial_number), date))
        .map(__to_tuple)
        # key by vault_id
        .key_by(lambda x: x[0])
    )

    window_1_day = (
        # create 1 day window
        ds.window(TumblingEventTimeWindows.of(Time.days(1)))
        # create reduction
        .reduce(__reduce)
        # create 1 day window for all the reduced elements
        .window_all(TumblingEventTimeWindows.of(Time.days(1)))
        # return the top 10 vaults (with correct csv formatting)
        .process(Top10Vaults())
    )
    window_3_days = (
        # create 1 day window
        ds.window(TumblingEventTimeWindows.of(Time.days(3)))
        # create reduction
        .reduce(__reduce)
        # create 3 days window for all the reduced elements
        .window_all(TumblingEventTimeWindows.of(Time.days(3)))
        # return the top 10 vaults (with correct csv formatting)
        .process(Top10Vaults())
    )
    window_all = (
        # create global window
        ds.window(GlobalWindows.create())
        # create reduction
        .reduce(__reduce)
        # create global window for all the reduced elements
        .window_all(GlobalWindows.create())
        # return the top 10 vaults (with correct csv formatting)
        .process(Top10Vaults())
    )

    return [
        (window_1_day, "query_2_day_1"),
        (window_3_days, "query_2_days_3"),
        (window_all, "query_2_days_all"),
    ]

File: src/flink/src/main/scala/queries/Query2.scala
package queries

import models.KafkaTuple
import models.QueryOutput
import models.QueryReturn
import org.apache.flink.api.common.functions.AggregateFunction
import org.apache.flink.streaming.api.datastream.DataStream
import org.apache.flink.streaming.api.datastream.KeyedStream
import org.apache.flink.streaming.api.functions.windowing.ProcessAllWindowFunction
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector

import java.time.Duration
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.math

object Query2 {
  private case class Accumulator(
      count: Long,
      ts: Long,
      vault_id: Int,
      failures: Int,
      hdds: mutable.Set[(String, String)] // Set((model, serial_number))
  )

  // Format csv output
  private class ProcessRanking
      extends ProcessAllWindowFunction[Accumulator, QueryOutput, TimeWindow] {
    override def process(
        context: ProcessAllWindowFunction[
          Accumulator,
          QueryOutput,
          TimeWindow
        ]#Context,
        elements: java.lang.Iterable[Accumulator],
        out: Collector[QueryOutput]
    ): Unit = {
      val sortedVaults =
        elements.asScala.toList.sortBy(-_.failures).take(10)

      var count = 0L
      var ts = Long.MinValue

      var result = DateTimeFormatter
        .ofPattern("yyyy-MM-dd")
        .withZone(ZoneOffset.UTC)
        .format(Instant.ofEpochMilli(context.window.getStart))
      for (vault <- sortedVaults) {
        result += s",${vault.vault_id},${vault.failures},("
        result += vault.hdds
          .map { case (model, serial_number) =>
            s"$model,$serial_number"
          }
          .mkString(",")
        result += ")"
        count += vault.count
        ts = math.max(ts, vault.ts)
      }
      out.collect(new QueryOutput(count, ts, result))
    }
  }

  // Compute failure count and list of failed hdds
  private class TupleAggregate
      extends AggregateFunction[KafkaTuple, Accumulator, Accumulator] {
    def createAccumulator(): Accumulator =
      new Accumulator(0, Long.MinValue, 0, 0, mutable.Set())

    def add(value: KafkaTuple, acc: Accumulator): Accumulator =
      Accumulator(
        acc.count + 1,
        math.max(value.ts, acc.ts),
        value.vault_id,
        acc.failures + value.failure,
        acc.hdds + ((value.model, value.serial_number))
      )

    def getResult(acc: Accumulator): Accumulator = acc

    def merge(a: Accumulator, b: Accumulator): Accumulator =
      Accumulator(
        a.count + b.count,
        math.max(a.ts, b.ts),
        a.vault_id,
        a.failures + b.failures,
        a.hdds ++ b.hdds
      )
  }

  def impl(
      ds: KeyedStream[KafkaTuple, Int],
      duration: Long,
      offset: Long
  ): DataStream[QueryOutput] = {
    return ds
      // Same as Query 1
      .window(
        TumblingEventTimeWindows
          .of(Duration.ofDays(duration), Duration.ofDays(offset))
      )
      // Compute aggregation
      .aggregate(new TupleAggregate)
      // compute ranking
      .windowAll(
        TumblingEventTimeWindows
          .of(Duration.ofDays(duration), Duration.ofDays(offset))
      )
      .process(new ProcessRanking)
  }

  def query(ds: DataStream[KafkaTuple]): List[QueryReturn] = {
    // Filter only tuples with a failure
    val working_ds = ds
      .filter(_.failure > 0)
      .keyBy(_.vault_id)

    return List(
      new QueryReturn(impl(working_ds, 1, 0), "query_1"),
      new QueryReturn(impl(working_ds, 3, 2), "query_3"),
      new QueryReturn(impl(working_ds, 23, 13), "query_23")
    )
  }
}
